File: SlyceSDKDemo/SlyceSDKDemo/MyCameraViewController.m
//
//  CustomCameraViewController.m
//  SlyceSDKDemo
//
//  Created by Nathan Gurfinkel on 2/3/15.
//  Copyright (c) 2015 Slyce. All rights reserved.
//

#import "MyCameraViewController.h"
#import "MyProductsViewController.h"
#import "MBProgressHUD.h"
#import <SlyceSDK/SlyceSDK.h>
#import <AVFoundation/AVFoundation.h>

//This is *your* custom view controller where you want to use the SDK's Camera view
@interface MyCameraViewController () <SFCameraViewDelegate, SFRequestDelegate, UIAlertViewDelegate>

@property (nonatomic, strong) SFCameraView *cameraView;

@property (nonatomic, copy) NSArray *products;
@property (nonatomic, strong) MBProgressHUD *hud;
@property (weak, nonatomic) IBOutlet UILabel *resultLabel;
@property (weak, nonatomic) IBOutlet UILabel *resultAdditionalLabel;
@property (weak, nonatomic) IBOutlet UILabel *premiumLabel;

@property (nonatomic, strong) SFRequest *currentRequest;
@property (nonatomic, strong) UIButton *cancelButton;

@end

@implementation MyCameraViewController

#pragma mark
#pragma mark - UIAlertViewDelegate

- (void)willPresentAlertView:(UIAlertView *)alertView
{
    [_cameraView pauseCapture];
}

- (void)alertView:(UIAlertView *)alertView didDismissWithButtonIndex:(NSInteger)buttonIndex
{
    [_cameraView resumeCapture];
}

#pragma mark -
#pragma mark Segues

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:@"ProductsSegue"])
    {
        MyProductsViewController *productsVC = segue.destinationViewController;
        productsVC.products = _products;
    }
}

#pragma mark -
#pragma mark SFCameraViewDelegate

- (void)sfCameraView:(SFCameraView *)cameraView didReceiveResults:(NSDictionary *)results
{
    NSArray *products = [results objectForKey:@"products"];
    
    NSLog(@"Recognized products are: %@", products);
    [self.hud hide:YES];
    _cancelButton.hidden = YES;
    
    self.products = [products copy];
    [self performSegueWithIdentifier:@"ProductsSegue" sender:nil];
}

- (void)sfCameraView:(SFCameraView *)cameraView didProgressToValue:(CGFloat)value withMessage:(NSString *)message
{
    NSLog(@"Finished %.2f percents, current step = %@", value, message);
    self.hud.labelText = message;
}

- (void)sfCameraView:(SFCameraView *)cameraView didFinishWithItemDescription:(NSDictionary *)itemDescription
{
    NSLog(@"Item description is: %@", itemDescription);
    [self.hud hide:YES];
    _cancelButton.hidden = YES;
}

- (void)sfCameraView:(SFCameraView *)cameraView didDetectBarcode:(SFBarcode *)barcode
{
    NSLog(@"Recognized Barcode Text = %@", barcode.text);
    [self.hud hide:YES];
    _cancelButton.hidden = YES;
    
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertView *alert = [[UIAlertView alloc] initWithTitle:barcode.typeString message:barcode.text delegate:self cancelButtonTitle:@"OK" otherButtonTitles:nil];
        [alert show];
    });
}

- (void)sfCameraView:(SFCameraView *)cameraView didFailWithError:(NSError *)error
{
    NSString *message = error.domain == SlyceErrorDomain ? [error sf_message] : [error localizedDescription];
    NSLog(@"Error = %@", message);
    
    [self.hud hide:YES];
    _cancelButton.hidden = YES;
    
    UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Error" message:message delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
    [alert show];
}

- (void)sfCameraView:(SFCameraView *)cameraView didReceiveImageInfo:(NSArray *)products
{
    NSLog(@"Recognized products are: %@", products);
    [self.hud hide:YES];
    _cancelButton.hidden = YES;
    
    self.products = [products copy];
    [self performSegueWithIdentifier:@"ProductsSegue" sender:nil];
}

- (void)sfCameraView:(SFCameraView *)cameraView didDetectImage:(NSDictionary *)imageInfo
{
    NSLog(@"Recognized 2D Products: data = %@", imageInfo);
    self.resultLabel.text = [imageInfo description];
}

- (void)sfCameraView:(SFCameraView *)cameraView didSnapImage:(UIImage *)image
{
    [self.hud show:YES];
    _cancelButton.hidden = NO;
}

- (void)sfCameraView:(SFCameraView *)cameraView wasTappedInPoint:(CGPoint)point
{
    NSLog(@"Tapped at Point = %@", NSStringFromCGPoint(point));
}

#pragma mark -
#pragma mark View LifeCycle

- (void)viewDidLayoutSubviews
{
    //Update the underlying video preview layer to match the frame
    if (self.cameraView)
        self.cameraView.previewLayer.frame = self.view.frame;
}

//Call `startCamera` of SFCamera object in `viewDidAppear:`
- (void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    
    //Start the camera ONLY after the view has actually appeared, as this operation is CPU consuming!
    [self.cameraView startCamera];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    NSString *isPremium = _slyce.isPremiumUser ? @"YES" : @"NO";
    NSString *is2DEnabled = _slyce.is2DSearchEnabled ? @"YES" : @"NO";
    self.premiumLabel.text = [NSString stringWithFormat:@"Premium = %@, 2D Enabled = %@", isPremium, is2DEnabled];
    
    self.hud = [[MBProgressHUD alloc] initWithView:self.view];
    self.hud.dimBackground = YES;
    self.hud.labelText = @"Sending image";
    self.hud.animationType = MBProgressHUDAnimationZoom;
    self.hud.mode = MBProgressHUDModeIndeterminate;
    [self.view addSubview:self.hud];
    
    [self addCancelButton];
    
    self.cameraView = [[SFCameraView alloc] initWithSlyce:_slyce view:self.view options:nil andDelegate:self];
    //self.cameraView.shouldUseContinuousRecognition = NO; //Uncomment this line if do not wish to get notified automatically about recognized barcodes and/or 2D items (Premium) (default is YES)
    //self.cameraView.shouldPauseScannerAfterRecognition = NO; //Uncomment this line if do not wish the auto scanner to pause after a successful detection (default is YES)
}

- (void)viewDidDisappear:(BOOL)animated
{
    [self.cameraView pauseCapture];
    [super viewDidDisappear:animated];
}

#pragma mark -
#pragma mark IBActions

- (void)cancelSearch
{
    [_currentRequest cancel];
    [self.hud hide:YES];
    _cancelButton.hidden = YES;
}

- (IBAction)back:(id)sender
{
    [self.navigationController popViewControllerAnimated:YES];
}

- (IBAction)snap:(id)sender
{
    self.hud.labelText = @"Sending image";
    [self.cameraView snap];
}

- (IBAction)toggleFlash:(UIButton *)sender
{
    sender.selected = !sender.selected;
    [self.cameraView turnFlash:sender.selected];
}

#pragma mark -
#pragma mark Helpers

- (void)addCancelButton
{
    self.cancelButton = [UIButton buttonWithType:UIButtonTypeCustom];
    [_cancelButton setTitleColor:[UIColor colorWithRed:192/255.f green:30/255.f blue:30/255.f alpha:1] forState:UIControlStateNormal];
    _cancelButton.frame = CGRectMake(self.hud.frame.size.width/2-70/2, self.hud.frame.size.height/2 + 60, 70, 30);
    _cancelButton.translatesAutoresizingMaskIntoConstraints = YES;
    [_cancelButton setTitle:@"Cancel" forState:UIControlStateNormal];
    _cancelButton.titleLabel.font = [UIFont systemFontOfSize:20.0];
    [_cancelButton addTarget:self action:@selector(cancelSearch) forControlEvents:UIControlEventTouchUpInside];
    _cancelButton.hidden = YES;
    [self.view addSubview:_cancelButton];
}

@end

File: SlyceSDKDemo_Swift/SlyceSDKSwift/MyCameraViewController.swift
//
//  MyCameraViewController.swift
//  SlyceSDKSwift
//
//  Created by HADI MRIH on 17/08/2016.
//  Copyright © 2016 Slyce inc. All rights reserved.
//

import UIKit

class MyCameraViewController: UIViewController,SFCameraViewDelegate,SFRequestDelegate {

    var cameraView = SFCameraView()
    var products = [AnyObject]()
    var hub = MBProgressHUD()
    @IBOutlet weak var cancelRewBtn: UIButton!
    
    override func viewDidDisappear(_ animated: Bool)
    {
        super.viewDidDisappear(animated)
        self.cameraView.pauseCapture()
    }
    
    override func viewDidLayoutSubviews()
    {
       self.cameraView.previewLayer.frame = self.view.frame
    }

    override func viewDidAppear(_ animated: Bool)
    {
        super.viewDidAppear(animated)
        
        self.cameraView.startCamera()
        self.cameraView.resumeCapture()
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        
        hub = MBProgressHUD.init(view: self.view)
        hub.dimBackground = true
        hub.labelText = "Sending image"
        hub.animationType = MBProgressHUDAnimationZoom
        hub.mode = MBProgressHUDModeIndeterminate
        self.view.addSubview(hub)
        
        
        self.cameraView = SFCameraView.init(slyce: SFSlyce.sharedInstance() as! SFSlyce, view: self.view, options: nil, andDelegate: self)
        
        self.view.bringSubview(toFront: cancelRewBtn)
        
     
        // Do any additional setup after loading the view.
    }
    
    
    
    @IBAction func snapBtnPressed(_ sender: AnyObject)
    {
        hub.labelText = "Sending image"

        self.cameraView.snap()
    }
    
    
    @IBAction func flipCameraPressed(_ sender: AnyObject)
    {
        cameraView.flipCameraPosition()
    }
    
    @IBAction func toggleFlashPressed(_ sender: AnyObject)
    {
        let toglleFlashBtn = sender as! UIButton
        toglleFlashBtn.isSelected = !toglleFlashBtn.isSelected
        cameraView.turnFlash(toglleFlashBtn.isSelected)
    }
    
    @IBAction func cancelReqBtnPressed(_ sender: AnyObject)
    {
        cameraView.resumeCapture()
        cameraView.currentRequest.cancel()
        hub.hide(true)
        cancelRewBtn.isHidden = true
    }
    
    // MARK: - SFCameraViewDelegate
    
    func sfCameraView(_ sfCameraView: SFCameraView!, didStartForImage image: UIImage!)
    {
        print("didStartForImage")
    }
    
    func sfCameraView(_ sfCameraView: SFCameraView!, didFinishWith statusType: SFStatusType)
    {
        print(NSString(format: "didFinishWithStatus: %@", String(describing: SFStatusType.self)));
    }
    
    func sfCameraView(_ sfCameraView: SFCameraView!, didDetect barcode: SFBarcode)
    {
        print(NSString(format: "didDetectBarcode: %@", barcode.text));
        
        cameraView.pauseCapture()

        let alertController = UIAlertController(title: barcode.typeString, message:
            barcode.text, preferredStyle: UIAlertControllerStyle.alert)
        
        alertController.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.default,handler:{(alert: UIAlertAction!) in self.cameraView.resumeCapture()}))
        
        self.present(alertController, animated: true, completion: nil)
    }
    
    func sfCameraView(_ sfCameraView: SFCameraView!, didReceiveResults results: [AnyHashable: Any]!)
    {
        print(NSString(format: "didReceiveResults:results %@",results));
        
        let resultsDic = results as NSDictionary
        
        let productsFromResults = resultsDic.object(forKey: "products") as! [AnyObject]

        self.products = productsFromResults
        
        self.performSegue(withIdentifier: "ProductsSegue", sender: nil)
        hub.hide(true)
        cancelRewBtn.isHidden = true

    }
    
    func sfCameraView(_ sfRequest: SFCameraView!, didFailWithError error: NSError!)
    {
        let message = error.domain == SlyceErrorDomain ? error.sf_message() : error.localizedDescription
        print(NSString(format: "didFailWithError:errorMessage %@", message!));
        
        let alertController = UIAlertController(title: "Error", message:
            message!, preferredStyle: UIAlertControllerStyle.alert)
        
        alertController.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.default,handler:{(alert: UIAlertAction!) in self.cameraView.resumeCapture()}))
        
        self.present(alertController, animated: true, completion: nil)
    }
    
    func sfCameraView(_ sfCameraView: SFCameraView!, didProgressToValue value: CGFloat, withMessage message: String!)
    {
        print(NSString(format: "didProgressToValue %f withMessage %@",value,message));
        
        hub.labelText = message
    }
    
    func sfCameraView(_ sfCameraView: SFCameraView!, didDetectImage imageInfo: [AnyHashable: Any]!)
    {
        print(NSString(format: "didDetectImage:imageInfo %@",imageInfo));
    }
    
    func sfCameraView(_ sfCameraView: SFCameraView!, didReceiveImageInfo products: [AnyObject]!)
    {
        print(NSString(format: "didReceiveImageInfo:products %@",products));
        self.products = products
        
        self.performSegue(withIdentifier: "ProductsSegue", sender: nil)
        hub.hide(true)
        cancelRewBtn.isHidden = true

    }
    
    func sfCameraView(_ sfCameraView: SFCameraView!, didReceiveResultsExt results: String)
    {
        print(NSString(format: "didReceiveResultsExt:results %@",results));
        hub.hide(true)
        cancelRewBtn.isHidden = true
    }
    
    func sfCameraView(_ sfCameraView: SFCameraView!, didProgressTo stage: SFRequestStage)
    {
        switch stage
        {
        case SFRequestStage.sendingImage:
            print("sfRequest:didProgressToStage:'Sending Image'")
            break
        case SFRequestStage.analyzingImage:
            print("sfRequest:didProgressToStage:'Analyzing Image'")
            break
        }
    }
    
    func sfCameraView(_ sfCameraView: SFCameraView!, didProgressExt progress: String)
    {
        print(NSString(format: "didProgressExt:progress %@",progress));
    }

    func sfCameraView(_ sfCameraView: SFCameraView!, didSnapImage image: UIImage)
    {
        hub.show(true)
        cameraView.pauseCapture()
        cancelRewBtn.isHidden = false

        print("sfCameraView:didSnapImage:")
    }
    
    func sfCameraView(_ sfCameraView: SFCameraView!, wasTappedIn point: CGPoint)
    {
        print(NSString(format: "sfCameraView:wasTappedInPoint: %@",NSStringFromCGPoint(point)));
    }
    

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        // Get the new view controller using segue.destinationViewController.
        // Pass the selected object to the new view controller.
        
        if segue.identifier == "ProductsSegue" {
            
            let productsVC  = segue.destination as! MyProductsViewController
            productsVC.products = self.products
        
        }
        
    }

}
